interactive, uat_decode, ais_charset: disambiguate unterminated strings

g7 opened the pull request:
GCC 15 adds -Wunterminated-string-initialization, which complains if a string is not NUL-terminated.

However, it causes false positives on byte arrays. Disambiguate that.

An alternative would be to use the [`nonstring` attribute](https://gcc.gnu.org/onlinedocs/gcc/Common-Variable-Attributes.html#index-nonstring-variable-attribute), which is available since GCC 8 and clang 21.
If you prefer using nonstring, I can adjust this Pull Request to change that.

wiedehopf commented:
Yeah nostring attribute seems preferrable.
Guess that would mean it fails for gcc 7 and clang 20 with that attribute?
Or is it simply ignored.

wiedehopf commented:
The downside of using char *var = "asdfsadf"; ?
Probably that sizeof var doesn't work anymore?

kasper93 commented:
> Guess that would mean it fails for gcc 7 and clang 20 with that attribute?
> Or is it simply ignored.

It would complain about unknown attribute. You can wrap with `__has_attribute` to resolve it.

But at least `ais_charset` is used as string with `strchr` in https://github.com/wiedehopf/readsb/blob/9cb37df8a96755cb70e79b1e189bcad2f892e353/uat2esnt/uat2esnt.c#L489-L496

and is duplicated there, so would likely make sens it to define it as null-terminated string globally. Quick grep didn't should any sizeof that would break after this change.

wiedehopf commented:
nostring attribute should be fine.
let's go witht he ifdef has_attribute just because so it compiles even on stupid old systems.

just copy pasta that for all 3 occurences if you want to.
otherwise i can do it as well.

g7 committed:
net_io: use global ais_charset definition

Signed-off-by: Eugenio Paolantonio (g7) <me@medesimo.eu>

g7 committed:
ais_charset, uat2esnt: ensure ais_charset is a NUL-terminated string

Signed-off-by: Eugenio Paolantonio (g7) <me@medesimo.eu>

g7 committed:
interactive, uat2esnt: use __nonstring__ attribute for byte arrays

GCC 15 adds -Wunterminated-string-initialization, which complains if a string
is not NUL-terminated.
However, it causes false positives on byte arrays.

Disambiguate that by using the __nonstring__ attribute, which is available
since GCC 8 and clang 21.

Signed-off-by: Eugenio Paolantonio (g7) <me@medesimo.eu>

g7 commented:
> But at least ais_charset is used as string with strchr in 

good catch. Looks like it is used in the same way in `net_io.c`.
I agree it makes sense to have it as a NUL-terminated string.

I have updated the Pull Request to do that, but I have kept the re-definition in `uat2esnt/uat2esnt.c` as if I understood correctly, uat2esnt was imported as a separate module and didn't want to add a dependency on a parent header there... 

Rest is using the `__nonstring__` attribute. Builds successfully in openSUSE Tumbleweed (GCC 15) and openSUSE Leap 15.6 (GCC 7).

Of course, if we want to keep ais_charset as a byte array rather than a string, I can add `__nonstring` there too.

wiedehopf commented:
don't think i care which way ais_charset goes.

seems fine either way.
maybe nicer as a string.

hmm just make everything a string maybe.
well might as well leave the nonstring where you already put it.

yeah i don't care, up to you.

just let me know when you have it in a state that doesn't complain in gcc15

wiedehopf commented:
So adding the 1 extra size on ais_charset fixed the error as well?
I guess because it's then a terminated string.
So the nostring thing isn't necessary if we just make the arrays 1 bigger?

I guess i'll merge this but i'm tempted to just throw it all over and make it all strings make it super simple.

g7 commented:
Yes, the current PR works fine in gcc15 (and also gcc7 at least, so hopefully it doesn't break other compilers).

> So adding the 1 extra size on ais_charset fixed the error as well?
> I guess because it's then a terminated string.
> So the nostring thing isn't necessary if we just make the arrays 1 bigger?

Yes. If you like it more this way, I can modify this PR again, having a skim at the code it shouldn't break stuff. And it will not need the __nostring boilerplate...

wiedehopf commented:
Could you check that dev compiles i've just added a commit that removes the nonstring attribute stuff in favor of just having the arrays 1 larger.

g7 commented:
Thank you! Tested with 21d399deb79e9c795e8e2ac35292e7f46becc6f9, it compiles fine with gcc 15.
